ui/ui.go

```diff
@@ -96,29 +96,52 @@ func (ui *WriterUI) PrintTable(table Table) {
 	}
 }
 
-func (ui *WriterUI) AskForText(label string) (string, error) {
+func (ui *WriterUI) AskForText(opts TextOpts) (string, error) {
 	var text string
 
-	err := interact.NewInteraction(label).Resolve(&text)
-	if err != nil {
-		return "", fmt.Errorf("Asking for text: %s", err)
+	for {
+		text = opts.Default
+		err := interact.NewInteraction(opts.Label).Resolve(&text)
+		if err != nil {
+			return "", fmt.Errorf("Asking for text: %s", err)
+		}
+
+		if opts.ValidateFunc != nil {
+			isValid, err := opts.ValidateFunc(text)
+			if isValid {
+				break
+			}
+			ui.PrintErrorBlock(err.Error())
+		}
 	}
 
 	return text, nil
 }
 
-func (ui *WriterUI) AskForChoice(label string, options []string) (int, error) {
-	var choices []interact.Choice
+func (ui *WriterUI) AskForChoice(opts ChoiceOpts) (int, error) {
+	var (
+		choices []interact.Choice
+		chosen  int
+	)
 
-	for i, opt := range options {
+	for i, opt := range opts.Choices {
 		choices = append(choices, interact.Choice{Display: opt, Value: i})
 	}
 
-	var chosen int
-
-	err := interact.NewInteraction(label, choices...).Resolve(&chosen)
-	if err != nil {
-		return 0, fmt.Errorf("Asking for choice: %s", err)
+	for {
+		chosen = opts.Default
```

Comment:
This will be outside for loop.

Reply:
Why I keep this inside the `for` loop is supposed in the first iteration instead of hitting enter user gave input on the terminal and suppose it's invalid then if we don't set the `chosen = opts.Default` again then the invalid input that the user-provided will become the Default now and I think we should not change our default in such scenarios.